### PR TITLE
Add MongoDb migration support to DataMigration service

### PR DESCRIPTION
**Pull request by haitsongmsft**
## Description

https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/207
## Checklist

- [ ] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by haitsongmsft** [comment]
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/207 is the issue PR.

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by maddieclayton** [comment]
@azuresdkci Add to whitelist

**Comment by haitsongmsft** [comment]
@cormacpayne,  My test shows field of "Parameter name: callerIdentity " is missing, per the build. I don't think we ever used this field, do you know what could be the cuase of this? https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/9272/artifact/src/Publish/TestResults/FailingTests/Microsoft.Azure.Commands.DataMigration.Test.dll.html is

**Comment by cormacpayne** [comment]
@haitsongmsft I re-targeted your PR at the new default branch, `master`, so you will need to pull from that branch to resolve the merge conflicts.

_Note_: since the next release is the GA of the `Az` module, we have removed all reference docs to `AzureRM` cmdlets, so you will need to regenerate your help based on the `Az` cmdlets.

**Comment by maddieclayton** [comment]
@azuresdkci Test this please

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewDataMigrationTask.cs [hunk 1]
```diff
@@ -176,12 +216,29 @@ public override void ExecuteCmdlet()
...
+                            Properties = properties
                         };
 
                         response = DataMigrationClient.Tasks.CreateOrUpdate(taskInput, ResourceGroupName, ServiceName, ProjectName, Name);
+
+                        // wait for the task to finish: not queued or running state:
+                        while (this.Wait.IsPresent && response !=null && response.Properties != null &&
+                            ( response.Properties.State == "Queued" || response.Properties.State == "Running" ) )
+                        {
+                            System.Threading.Thread.Sleep(5000);
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewDataMigrationTask.cs]: @haitsongmsft nit: I'd replace `5000` with `TimeSpan.FromSeconds(5))`, as well as replacing the `5` with a variable at the top of this file that defines the value.

File: src/ResourceManager/DataMigration/Commands.DataMigration/Models/MongoDb/MongoDbCollectionSetting.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Microsoft.Azure.Commands.DataMigration.Models.MongoDb
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/Models/MongoDb/MongoDbCollectionSetting.cs]: @haitsongmsft please make sure to add the Microsoft license header to all `.cs` files that are missing it

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    [Cmdlet(VerbsCommon.New, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "DataMigrationMongoDbCollectionSetting", SupportsShouldProcess = true), OutputType(typeof(MongoDbCollectionSetting))]
+    [Alias("New-"+ ResourceManager.Common.AzureRMConstants.AzureRMPrefix +"DmsMongoDbCollectionSetting")]
+    public class NewMongoDbCollectionSettingCmdlet : DataMigrationCmdlet
+    {
+        [Parameter(
+            Mandatory = true,
+            HelpMessage = "The name of the database collection.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("Name")]
+        public string CollectionName;
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs]: @haitsongmsft I would flip `Name` and `CollectionName` here (`Name` is usually used for `New-*` cmdlet)

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        [Parameter(
+           Mandatory = false,
+           HelpMessage = "The database level shared RU at the target CosmosDb"
+               )]
+        [Alias("RU")]
+        public int? TargetRequestUnit { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs]: @haitsongmsft is there a specific range of values that the user can provide? If so, I'd recommend using the `ValidateRange` attribute to limit the integers that a user could provide to this parameter.

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    [Cmdlet(VerbsCommon.New, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "DataMigrationMongoDbDatabaseSetting", SupportsShouldProcess = true), OutputType(typeof(MongoDbDatabaseSetting))]
+    [Alias("New-" + ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "DmsMongoDbDatabaseSetting")]
+    public class NewMongoDbDatabaseSettingCmdlet : DataMigrationCmdlet
+    {
+        [Parameter(
+           Mandatory = true,
+           HelpMessage = "The mongo db database name"
+               )]
+        [Alias("Name")]
+        public string DatabaseName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs]: @haitsongmsft same comment here about swapping `Name` and `DatabaseName`

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+        [Parameter(
+           Mandatory = false,
+           HelpMessage = "The database level shared RU at the target CosmosDb"
+               )]
+        [Alias("RU")]
+        public int? TargetRequestUnit { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs]: @haitsongmsft same comment here about potentially adding a `ValidateRange` attribute

File: src/ResourceManager/DataMigration/Commands.DataMigration/help/New-AzDataMigrationMongoDbCollectionSetting.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+## DESCRIPTION
+The New-AzDataMigrationMongoDbCollectionSetting cmdlet creates the migration setting object that specifies the throughput and delete behavior.
+The output the cmdlet is key value pair with name of the collection, and value of the setting. The output is used in assembling
+the database level settings for migration.
+
+## EXAMPLES
+
+### Example 1
+```
+PS C:\> New-AzDataMigrationMongoDbCollectionSetting -Name myCollection -TargetRequestUnit 1000 -CanDelete -ShardKey "_id:-1,age:1,name"
```
Comment (cormacpayne) [inline src/ResourceManager/DataMigration/Commands.DataMigration/help/New-AzDataMigrationMongoDbCollectionSetting.md]: @haitsongmsft for the examples in the markdown files you're adding, would you mind providing the resulting output if the user were to run the command? This gives users a better idea of what properties can be found on the object returned from the command

**Comment by cormacpayne** [comment]
@haitsongmsft ping on the comments above

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewDataMigrationTask.cs (hunk 1 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewDataMigrationTask.cs]: updated

File: src/ResourceManager/DataMigration/Commands.DataMigration/Models/MongoDb/MongoDbCollectionSetting.cs (hunk 2 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/Models/MongoDb/MongoDbCollectionSetting.cs]: updated

File: src/ResourceManager/DataMigration/Commands.DataMigration/AzureRM.DataMigration.psd1 [hunk 8]
```diff
@@ -87,6 +87,8 @@ CmdletsToExport = 'New-AzureRmDataMigrationDatabaseInfo',
                'New-AzureRmDataMigrationSelectedDBObject', 
                'New-AzureRmDataMigrationFileShare', 
                'New-AzureRmDataMigrationSyncSelectedDBObject', 
+               'New-AzureRmDataMigrationMongoDbDatabaseSetting',
```
[the remark this answers is not in the document]
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/AzureRM.DataMigration.psd1]: updated

File: src/ResourceManager/DataMigration/Commands.DataMigration/help/New-AzDataMigrationMongoDbCollectionSetting.md (hunk 7 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/help/New-AzDataMigrationMongoDbCollectionSetting.md]: updated

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs (hunk 6 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs]: updated.

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs (hunk 5 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbDatabaseSettingCmdlet.cs]: updated.

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs (hunk 4 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs]: Good catch, the range is 400-1000000

File: src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs (hunk 3 above)
Comment (haitsongmsft) [reply src/ResourceManager/DataMigration/Commands.DataMigration/Cmdlets/NewMongoDbCollectionSettingCmdlet.cs]: updated

**Comment by cormacpayne** [comment]
@haitsongmsft would you mind resolving the merge conflicts?

**Pull request merged**